packages/editor/src/components/post-author/enhanced-post-author.js

```diff
@@ -0,0 +1,115 @@
+/**
+ * External dependencies
+ */
+import { debounce } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useState } from '@wordpress/element';
+import apiFetch from '@wordpress/api-fetch';
+
+/**
+ * Internal dependencies
+ */
+import ComboboxControl from '../../../../components/build/combobox-control/';
+
+function EnhancedPostAuthor( { postAuthor, id, authors, onUpdateAuthor } ) {
+	const currentPostAuthor = postAuthor[ 0 ];
+	let initialAuthors = [];
+	authors.forEach( ( author, i ) => {
+		initialAuthors = [
+			...initialAuthors,
+			{
+				key: i,
+				name: author.name,
+				value: author.name,
+				id: author.id,
+			},
+		];
+	} );
+
+	// Ensure the current author is included in the initial dropdown list.
+	let foundAuthor = initialAuthors.findIndex(
+		( author ) => currentPostAuthor && currentPostAuthor.id === author.id
+	);
+
+	if ( currentPostAuthor && foundAuthor < 0 ) {
+		currentPostAuthor.key = initialAuthors.length;
+		initialAuthors = [ currentPostAuthor, ...initialAuthors ];
+		foundAuthor = 0;
+	}
+
+	// The search result loading state.
+	const [ isLoadingSearchResults, setIsLoadingSearchResult ] = useState(
+		false
+	);
+
+	// The list of authors available in the dropdown.
+	const [ availableAuthors, setAvailableAuthors ] = useState(
+		initialAuthors
+	);
+
+	// The currently selected author.
+	const [ selectedAuthor, setSelectedAuthor ] = useState( currentPostAuthor );
+
+	/**
+	 * Handle author selection.
+	 *
+	 * @param {Object} selectedItem The selected Author.
+	 */
+	const handleSelect = ( { selectedItem } ) => {
+		setSelectedAuthor( selectedItem );
+		onUpdateAuthor( selectedItem.id );
+	};
+
+	/**
+	 * Handle user input.
+	 *
+	 * @param {string} inputValue The current value of the input field.
+	 */
+	const handleKeydown = ( { inputValue } ) => {
+		if ( '' === inputValue ) {
+			// When the field is cleared, reset the author list to the original list.
+			setAvailableAuthors( initialAuthors );
+		} else {
+			// Otherwise search for authors matching the input.
+			searchAuthors( inputValue );
+		}
+	};
+
+	/**
+	 * Search for authors from the rest API.
+	 *
+	 * @param string query The string to search for.
+	 **/
+	const searchAuthors = debounce( ( query ) => {
+		const payload = '?search=' + encodeURIComponent( query );
+		setIsLoadingSearchResult( true );
+		apiFetch( { path: '/wp/v2/users' + payload } ).then( ( results ) => {
```

Comment:
Doesn't this need `who=authors`?

Reply:
This also should be done in `useSelect`, `getAuthors` is built in a way that supports adding queries.

Reply:
I'll give using getAuthors a try here, I didn't know we could pass a query. Not sure what `who=authors` does I will review.

Reply:
Not seeing how `getAuthors` support adding queries? I will add that.

Reply:
I added a new searchAuthors selector and resolver, if we use the original getAuthors, it changes the authors property which we don't want.

I switched to `useSelect` which simplifies the code here. I'm no longer tracking loading state because of this change. Should I add something to the resolver to yield a 'SEARCHING' state until it completes... or any suggestions for how to track a loading state while the resolver is fetching the data?

Reply:
We don't need a `searchAuthors`, `getAuthors` needs a new query parameter that works like `getEntityRecords`'s.

Reply:
I tried this approach and can re-visit; with the current setup, I found getAuthors with parameter would overwrite original state authors, causing a prop change.

How about tracking the loading state question?

Reply:
`receiveUserQuery`'s first parameter is a query key.

See how entity record queries use packages/core-data/src/queried-data/get-query-parts.js to generate keys from the query.

>How about tracking the loading state question?

You should be able to use `select( 'core/data' ).isResolving`.

Reply:
> See how entity record queries use packages/core-data/src/queried-data/get-query-parts.js to generate keys from the query.

Excellent, thanks for the pointer.

> You should be able to use select( 'core/data' ).isResolving.

Brilliant, I'll try that.

Reply:
I still found I needed to construct a unique key, which is maybe fine - this accounts for an empty query.

Reply:
Added.